### PR TITLE
Drop intermediate ID set in `TransportGetSnapshotsAction`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we build an `Iterator<SnapshotId>` to compute the snapshot IDs
whose `SnapshotInfo` should be returned. This iterator first looks at
ongoing snapshots and then the completed ones. When consuming items from
this iterator we check again whether the snapshot ID corresponds to an
ongoing snapshot and produce the corresponding `SnapshotInfo`
differently in that case. This forces us to create a potentially-massive
intermediate set of all the `SnapshotId` values we need to look up.

There's no need for these two phases, we can instead build the two
different ways of creating a `SnapshotInfo` into the iterator itself,
saving the large intermediate set of IDs and simplifying the
implementation.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 1]
```diff
@@ -328,88 +328,105 @@ private void ensureRequiredNamesPresent(String repositoryName, @Nullable Reposit
...
-         * @return an iterator over the snapshot IDs in the given repository which match {@link #snapshotNamePredicate}.
+         * An asynchronous supplier of a {@link SnapshotInfo}.
          */
-        private Iterator<SnapshotId> getSnapshotIdIterator(String repositoryName, @Nullable RepositoryData repositoryData) {
+        private interface AsyncSnapshotInfo {
+
+            /**
+             * @return the {@link SnapshotId} of the {@link SnapshotInfo} to be retrieved.
+             */
+            SnapshotId getSnapshotId();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:338]: There'll eventually be no need for this, once we handle `addSimpleSnapshotInfos` in the same way, but I'm leaving it like this for now to keep the PR focussed.

**Comment by mhl-b** [comment]
Does it replace intermediate set of SnapshotId with same size of AsyncSnapshotInfo? I read code few times but still cant tell what is the gain.

**Comment by DaveCTurner** [comment]
> Does it replace intermediate set of SnapshotId with same size of AsyncSnapshotInfo?

There's no intermediate set any more (on the `verbose == true` path anyway).

**Review by mhl-b — APPROVED** [review]
L-kind-of-GTM

**Pull request merged**